src/components/PivotTable/PivotTableGrid/util.ts

```diff
@@ -0,0 +1,579 @@
+  const isMixedTable = !isNil(mixedTable)
+
+  buildVerticalTableHeader<T>(keys, rowHeaderSpace, divs, keysMapping)
+
+  /**
+   * Build values and keys
+   */
+  for (let cell of cellData) {
+    if (isMixedTable) {
+      if (cell.key === mixedTable.totalKey) {
+        const rowTotalValuesKey: string = cell.path.replace(
+          getCurrentPath(cell.key.toString(), cell.cellValue.toString()),
+          ''
+        )
+        rowTotalValues.set(rowTotalValuesKey, cell.total || 0)
+      }
+      if (!keys.includes(cell.key)) {
+        continue
+      }
+    }
+    const lastKey = mixedTable && cell.key === keys[keys.length - 1]
+    const { cellValue: value, cellSpan: span, column: columnStart = 0 } = cell
+    const columnEnd = lastKey ? columnStart + 2 : columnStart + 1
+    const rowStart = getCellInitialPosition(cell.initialPosition) + rowHeaderSpace
+    const rowEnd = rowStart + span.value
+    maxRowEnd = max([rowEnd, maxRowEnd])
+    maxColumnEnd = columnEnd > maxColumnEnd ? columnStart : maxColumnEnd
+
+    const valuesGridArea = new GridArea(rowStart, columnStart, rowEnd, columnEnd)
+    const formattedValue = typeof value === 'number' ? formatDecimalOrInteger(value) : value
+
+    divs.push({
+      gridArea: valuesGridArea,
+      types: new Set(cell.key !== RESULT_PATH_KEY ? [PivotTableCellType.HEADER] : [PivotTableCellType.VALUE]),
+      children: formattedValue,
+      isEndColumn: cell.key === RESULT_PATH_KEY && !mixedTable,
+    })
+
+    cellPosition.add(valuesGridArea.toString())
+  }
+
+  buildVerticalTableTotals<T>(
+    isMixedTable,
+    maxRowEnd,
+    keys,
+    rowHeaderSpace,
+    maxColumnEnd,
+    divs,
+    tree.nodeValue,
+    cellPosition
+  )
+
+  return { divs, rowTotalValues, totalRowNumber: maxRowEnd, cellPosition }
+}
+
+function getHorizontal<T extends object>(params: HorizontalTableProps<T>): PivotTableCellProps[] {
+  const { cellData, keys, tree, keysMapping, columnHeaderSpace = 1, mixedTable } = params
+  const divs: PivotTableCellProps[] = []
+  const mixedTableStartRowCache = new Map<string, number>()
+  const mixedTableColumnTotals = new Map<number, number>()
+  const cellPositions = mixedTable?.cellPosition || new Set<string>()
+
+  buildHorizontalTableHeader<T>(keys, columnHeaderSpace, divs, keysMapping)
+
+  const { maxRowEnd, maxColumnEnd } = buildValuesAndKeys<T>(
+    cellData,
+    columnHeaderSpace,
+    mixedTable,
+    mixedTableColumnTotals,
+    keys,
+    mixedTableStartRowCache,
+    divs,
+    cellPositions
+  )
+
+  if (mixedTable) {
+    buildMixedTableTotals<T>(
+      maxColumnEnd,
+      keys,
+      mixedTable.rowTotalValues,
+      mixedTableColumnTotals,
+      mixedTable.totalRowNumber,
+      divs,
+      cellPositions,
+      mixedTableStartRowCache,
+      columnHeaderSpace,
+      tree.nodeValue
+    )
+  } else {
+    buildHorizontalTableTotals(maxRowEnd, maxColumnEnd, divs, tree.nodeValue)
+  }
+  return divs
+}
+
+function buildValuesAndKeys<T extends object>(
```

Comment:
why is it called buildValuesAndKeys and returns `{ maxRowEnd, maxColumnEnd }`. :thinking:

Reply:
It was altering the parameter objects, I'll change it so it returns a list.